File: objective-c-sample/JSQMessagesViewController_SendBird/Messaging/ViewController/UserListViewController.m
//
//  MemberListViewController.m
//  JSQMessagesViewController_SendBird
//
//  Created by Jed Gyeong on 4/25/16.
//  Copyright © 2016 SENDBIRD. All rights reserved.
//

#import "UserListViewController.h"
#import "MemberListTableViewCell.h"
#import "MessagingViewController.h"

@interface UserListViewController ()

@property (weak, nonatomic) IBOutlet UITableView *memberListTableView;

@property (strong, nonatomic) NSMutableArray<SBDUser *> *members;
@property (atomic) BOOL isLoading;

@property (atomic) NSString *userID;
@property (atomic) NSString *userName;

@property (strong, nonatomic) SBDUserListQuery *memberListQuery;
@property (strong, nonatomic) NSMutableDictionary<NSString *, NSNumber *> *selectedMembers;

@end

@implementation UserListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    self.title = @"User List";
    
    self.isLoading = NO;
    self.members = [[NSMutableArray alloc] init];
    self.selectedMembers = [[NSMutableDictionary alloc] init];
    
    self.memberListTableView.delegate = self;
    self.memberListTableView.dataSource = self;
    [self.memberListTableView registerNib:[MemberListTableViewCell nib] forCellReuseIdentifier:[MemberListTableViewCell cellReuseIdentifier]];
    
    self.memberListQuery = [SBDMain createAllUserListQuery];
    
    [self loadMembers];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];

    if (self.invitationMode == 0) {
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Create Channel" style:UIBarButtonItemStylePlain target:self action:@selector(clickCreateChannel:)];
    }
    else {
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Invite" style:UIBarButtonItemStylePlain target:self action:@selector(clickInvite:)];
    }
}

- (void)loadMembers {
    if (self.isLoading) {
        return;
    }
    
    if (!self.memberListQuery.hasNext) {
        return;
    }
    
    self.isLoading = YES;
    
    [self.memberListQuery loadNextPageWithCompletionHandler:^(NSArray<SBDUser *> * _Nullable users, SBDError * _Nullable error) {
        if (error != nil) {
            self.isLoading = NO;
            
            return;
        }
        
        if (users == nil || users.count == 0) {
            self.isLoading = NO;
            
            return;
        }
        
        for (SBDUser *user in users) {
            if ([user.userId isEqualToString:[SBDMain getCurrentUser].userId]) {
                continue;
            }
            [self.members addObject:user];
            self.selectedMembers[user.userId] = @(0);
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.memberListTableView reloadData];
        });

        self.isLoading = NO;
    }];
}

- (void)clickCreateChannel:(id)sender {
    NSMutableArray<NSString *> *userIds = [[NSMutableArray alloc] init];
    
    for (NSString *key in self.selectedMembers) {
        if ([[self.selectedMembers objectForKey:key] intValue] == 1) {
            [userIds addObject:key];
        }
    }
    
    if ([userIds count] > 0) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Create Group Channel" message:@"Create a group channel." preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *closeAction = [UIAlertAction actionWithTitle:@"Close" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            
        }];
        
        UIAlertAction *createDistinctChannelAction = [UIAlertAction actionWithTitle:@"Create distinct channel" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [SBDGroupChannel createChannelWithUserIds:userIds isDistinct:YES completionHandler:^(SBDGroupChannel * _Nullable channel, SBDError * _Nullable error) {
                if (error != nil) {
                    return;
                }
                
                if (self.delegate != nil && [self.delegate respondsToSelector:@selector(didCloseUserListViewController:groupChannel:)]) {
                    [self.delegate didCloseUserListViewController:self groupChannel:channel];
                }
                
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.navigationController popViewControllerAnimated:NO];
                });
            }];
        }];
        
        UIAlertAction *createNonDistinctChannelAction = [UIAlertAction actionWithTitle:@"Create non-distinct channel" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [SBDGroupChannel createChannelWithUserIds:userIds isDistinct:NO completionHandler:^(SBDGroupChannel * _Nullable channel, SBDError * _Nullable error) {
                if (error != nil) {
                    return;
                }
                
                if (self.delegate != nil && [self.delegate respondsToSelector:@selector(didCloseUserListViewController:groupChannel:)]) {
                    [self.delegate didCloseUserListViewController:self groupChannel:channel];
                }
                
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.navigationController popViewControllerAnimated:NO];
                });
            }];
        }];

        [alert addAction:closeAction];
        [alert addAction:createDistinctChannelAction];
        [alert addAction:createNonDistinctChannelAction];
        
        [self presentViewController:alert animated:YES completion:nil];
    }
    else {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Select Users" message:@"You have to select users to include" preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *closeAction = [UIAlertAction actionWithTitle:@"Close" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            
        }];
        
        [alert addAction:closeAction];
        [self presentViewController:alert animated:YES completion:nil];
    }
}

- (void)clickInvite:(id)sender {
    NSMutableArray<NSString *> *userIds = [[NSMutableArray alloc] init];
    
    for (NSString *key in self.selectedMembers) {
        if ([[self.selectedMembers objectForKey:key] intValue] == 1) {
            [userIds addObject:key];
        }
    }

    if ([userIds count] > 0) {
        [self.currentChannel inviteUserIds:userIds completionHandler:^(SBDError * _Nullable error) {
            if (error != nil) {
                NSLog(@"Error: %@", error);
            }
            
            dispatch_async(dispatch_get_main_queue(), ^{
                if (self.delegate != nil && [self.delegate respondsToSelector:@selector(didCloseUserListViewController:groupChannel:)]) {
                    [self.delegate didCloseUserListViewController:self groupChannel:self.currentChannel];
                }
                
                [self.navigationController popViewControllerAnimated:YES];
            });
        }];
    }
}

#pragma mark - UITableViewDelegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 48;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 48;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForHeaderInSection:(NSInteger)section
{
    return 0;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForFooterInSection:(NSInteger)section
{
    return 0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    SBDUser *user = [self.members objectAtIndex:indexPath.row];
    UITableViewCell* cellCheck = [tableView cellForRowAtIndexPath:indexPath];
    
    if ([[self.selectedMembers objectForKey:user.userId] intValue] == 0) {
        [cellCheck setAccessoryType:UITableViewCellAccessoryCheckmark];
        [self.selectedMembers setObject:@(1) forKey:user.userId];
    }
    else {
        [cellCheck setAccessoryType:UITableViewCellAccessoryNone];
        [self.selectedMembers setObject:@(0) forKey:user.userId];
    }
    
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}

#pragma mark - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.members count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    SBDUser *user = [self.members objectAtIndex:indexPath.row];
    MemberListTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:[MemberListTableViewCell cellReuseIdentifier]];
    
    if (cell == nil) {
        cell = [[MemberListTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[MemberListTableViewCell cellReuseIdentifier]];
    }
    
    [cell setModel:user];
    [cell setOnlineStatusVisiblility:YES];
    
    if ([[self.selectedMembers objectForKey:user.userId] intValue] == 1) {
        [cell setAccessoryType:UITableViewCellAccessoryCheckmark];
    }
    else {
        [cell setAccessoryType:UITableViewCellAccessoryNone];
    }
    
    if ([self.members count] > 0) {
        if (indexPath.row == [self.members count] - 1) {
            [self loadMembers];
        }
    }

    return cell;
}

@end

File: swift-3-sample/SampleUI/UserListViewController.swift
//
//  UserListViewController.swift
//  SampleUI
//
//  Created by Jed Kyung on 8/24/16.
//  Copyright © 2016 SendBird. All rights reserved.
//

import UIKit
import SendBirdSDK

protocol UserListViewControllerDelegate: class {
    func didCloseUserListViewController(_ vc: UIViewController, groupChannel: SBDGroupChannel)
}

class UserListViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    weak var delegate: UserListViewControllerDelegate!
    var invitationMode: Int?
    var channel: SBDGroupChannel?

    @IBOutlet fileprivate weak var tableView: UITableView!
    fileprivate var users: [SBDUser] = []
    fileprivate var userId: String?
    fileprivate var userName: String?
    
    fileprivate var userListQuery: SBDUserListQuery?
    fileprivate var selectedUsers: NSMutableDictionary?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.title = "User List"
        
        self.selectedUsers = NSMutableDictionary()
        
        self.tableView.delegate = self
        self.tableView.dataSource = self
        self.tableView.register(UserListTableViewCell.nib(), forCellReuseIdentifier: UserListTableViewCell.cellReuseIdentifier())
        
        self.userListQuery = SBDMain.createAllUserListQuery()
        self.loadUsers()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if self.invitationMode == 0 {
            self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Create Channel", style: UIBarButtonItemStyle.plain, target: self, action: #selector(UserListViewController.clickCreateChannel(_:)))
        }
        else {
            self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Invite", style: UIBarButtonItemStyle.plain, target: self, action: #selector(UserListViewController.clickInvite(_:)))
        }
    }
    
    fileprivate func loadUsers() {
        if self.userListQuery?.isLoading() == true {
            return
        }
        
        if self.userListQuery?.hasNext == false {
            return
        }
        
        self.userListQuery?.loadNextPage(completionHandler: { (users, error) in
            if error != nil {
                return
            }
            
            if users == nil || users!.count == 0 {
                return
            }
            
            for item in users! {
                let user = item 
                if user.userId == SBDMain.getCurrentUser()!.userId {
                    continue
                }
                self.users.append(user)
                self.selectedUsers![user.userId] = Int(0)
            }
            
            DispatchQueue.main.async(execute: { 
                self.tableView.reloadData()
            })
        })
    }
    
    func clickCreateChannel(_ sender: AnyObject) {
        var userIds = [String]()
        
        for item in self.selectedUsers! {
            if item.value as! Int == 1 {
                userIds.append(item.key as! String)
            }
        }
        
        if userIds.count > 0 {
            let alert = UIAlertController(title: "Create Group Channel", message: "Create a group channel.", preferredStyle: UIAlertControllerStyle.alert)
            let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
            let createDistinctChannelAction = UIAlertAction(title: "Create distinct channel", style: UIAlertActionStyle.default, handler: { (action) in
                SBDGroupChannel.createChannel(withUserIds: userIds, isDistinct: true, completionHandler: { (channel, error) in
                    if error != nil {
                        let alert = UIAlertController(title: "Error", message: String(format: "%lld: %@", error!.code, (error?.domain)!), preferredStyle: UIAlertControllerStyle.alert)
                        let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
                        alert.addAction(closeAction)
                        DispatchQueue.main.async(execute: {
                            self.present(alert, animated: true, completion: nil)
                        })
                        
                        return
                    }
                    
                    if self.delegate != nil {
                        self.delegate?.didCloseUserListViewController(self, groupChannel: channel!)
                    }
                    
                    DispatchQueue.main.async(execute: { 
                        self.navigationController!.popViewController(animated: false)
                    })
                })
            })
            let createNonDistinctChannelAction = UIAlertAction(title: "Create non-distinct channel", style: UIAlertActionStyle.default, handler: { (action) in
                SBDGroupChannel.createChannel(withUserIds: userIds, isDistinct: false, completionHandler: { (channel, error) in
                    if error != nil {
                        let alert = UIAlertController(title: "Error", message: String(format: "%lld: %@", error!.code, (error?.domain)!), preferredStyle: UIAlertControllerStyle.alert)
                        let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
                        alert.addAction(closeAction)
                        DispatchQueue.main.async(execute: {
                            self.present(alert, animated: true, completion: nil)
                        })
                        
                        return
                    }
                    
                    if self.delegate != nil {
                        self.delegate?.didCloseUserListViewController(self, groupChannel: channel!)
                    }
                    
                    DispatchQueue.main.async(execute: {
                        self.navigationController!.popViewController(animated: false)
                    })
                })
            })
            
            alert.addAction(closeAction)
            alert.addAction(createDistinctChannelAction)
            alert.addAction(createNonDistinctChannelAction)
            
            self.present(alert, animated: true, completion: nil)
        }
        else {
            let alert = UIAlertController(title: "Select Users", message: "You have to select users to include", preferredStyle: UIAlertControllerStyle.alert)
            let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
            
            alert.addAction(closeAction)
            self.present(alert, animated: true, completion: nil)
        }
    }
    
    func clickInvite(_ sender: AnyObject) {
        var userIds = [String]()
        
        for item in self.selectedUsers! {
            if item.value as! Int == 1 {
                userIds.append(item.key as! String)
            }
        }
        
        if userIds.count > 0 {
            self.channel?.inviteUserIds(userIds, completionHandler: { (error) in
                if error != nil {
                    let alert = UIAlertController(title: "Error", message: String(format: "%lld: %@", error!.code, (error?.domain)!), preferredStyle: UIAlertControllerStyle.alert)
                    let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
                    alert.addAction(closeAction)
                    DispatchQueue.main.async(execute: {
                        self.present(alert, animated: true, completion: nil)
                    })
                    
                    return
                }
                
                DispatchQueue.main.async(execute: { 
                    if self.delegate != nil {
                        self.delegate?.didCloseUserListViewController(self, groupChannel: self.channel!)
                    }
                    
                    self.navigationController!.popViewController(animated: true)
                })
            })
        }
    }
    
    // MARK: UITableViewDelegate
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 48
    }
    
    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return 48
    }
    
    func tableView(_ tableView: UITableView, estimatedHeightForFooterInSection section: Int) -> CGFloat {
        return 0
    }
    
    func tableView(_ tableView: UITableView, estimatedHeightForHeaderInSection section: Int) -> CGFloat {
        return 0
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let user = self.users[(indexPath as NSIndexPath).row]
        let cellCheck = tableView.cellForRow(at: indexPath)
        
        if (self.selectedUsers?.object(forKey: user.userId) as AnyObject).int32Value == 0 {
            cellCheck?.accessoryType = UITableViewCellAccessoryType.checkmark
            self.selectedUsers?.setObject(Int(1), forKey: user.userId as NSCopying)
        }
        else {
            cellCheck?.accessoryType = UITableViewCellAccessoryType.none
            self.selectedUsers?.setObject(Int(0), forKey: user.userId as NSCopying)
        }
        
        tableView.deselectRow(at: indexPath, animated: false)
    }
    
    // MARK: UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.users.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let user = self.users[(indexPath as NSIndexPath).row]
        let cell = tableView.dequeueReusableCell(withIdentifier: UserListTableViewCell.cellReuseIdentifier()) as! UserListTableViewCell
        
        cell.setModel(user)
        cell.setOnlineStatusVisiblility(true)
        
        if (self.selectedUsers?.object(forKey: user.userId) as AnyObject).int32Value == 1 {
            cell.accessoryType = UITableViewCellAccessoryType.checkmark
        }
        else {
            cell.accessoryType = UITableViewCellAccessoryType.none
        }
        
        if self.users.count > 0 {
            if (indexPath as NSIndexPath).row == self.users.count - 1 {
                self.loadUsers()
            }
        }
        
        return cell
    }
}
